block/sequencers.go

```diff
@@ -0,0 +1,198 @@
+	if m.SLClient.GetProposer() != nil {
+		expectedHubProposer = m.SLClient.GetProposer().PublicKey.Bytes()
+	}
+	return bytes.Equal(l2Proposer, localProposerKey) || bytes.Equal(expectedHubProposer, localProposerKey)
+}
+
+// check rotation in progress (I'm the proposer, but needs to complete rotation)
+func (m *Manager) MissingLastBatch() (string, error) {
+	localProposerKey, _ := m.LocalKey.GetPublic().Raw()
+	next, err := m.SLClient.CheckRotationInProgress()
+	if err != nil {
+		return "", err
+	}
+	if next == nil {
+		return "", nil
+	}
+	// rotation in progress,
+	// check if we're the old proposer and needs to complete rotation
+	if !bytes.Equal(next.PublicKey.Bytes(), localProposerKey) {
+		return next.Address, nil
+	}
+
+	return "", nil
+}
+
+// handleRotationReq completes the rotation flow once a signal is received from the SL
+// this called after manager shuts down the block producer and submitter
+func (m *Manager) handleRotationReq(ctx context.Context, nextSeqAddr string) {
+	m.logger.Info("Sequencer rotation started. Production stopped on this sequencer", "nextSeqAddr", nextSeqAddr)
+	err := m.CompleteRotation(ctx, nextSeqAddr)
+	if err != nil {
+		panic(err)
+	}
+
+	// TODO: graceful fallback to full node (https://github.com/dymensionxyz/dymint/issues/1008)
+	m.logger.Info("Sequencer is no longer the proposer")
+	panic("sequencer is no longer the proposer")
+}
+
+// CompleteRotation completes the sequencer rotation flow
+// the sequencer will create his last block, with the next sequencer's hash, to handover the proposer role
+// then it will submit all the data accumulated thus far and mark the last state update
+// if nextSeqAddr is empty, the nodes will halt after applying the block produced
+func (m *Manager) CompleteRotation(ctx context.Context, nextSeqAddr string) error {
+	// validate nextSeq is in the bonded set
+	var nextSeqHash [32]byte
+	if nextSeqAddr != "" {
+		val := m.State.Sequencers.GetByAddress([]byte(nextSeqAddr))
+		if val == nil {
+			return types.ErrMissingProposerPubKey
+		}
+		copy(nextSeqHash[:], val.PubKey.Address().Bytes())
+	}
+
+	err := m.CreateAndPostLastBatch(ctx, nextSeqHash)
+	if err != nil {
+		return fmt.Errorf("create and post last batch: %w", err)
+	}
+
+	m.logger.Info("Sequencer rotation completed. sequencer is no longer the proposer", "nextSeqAddr", nextSeqAddr)
+	return nil
+}
+
+// CreateAndPostLastBatch creates and posts the last batch to the hub
+// this called after manager shuts down the block producer and submitter
+func (m *Manager) CreateAndPostLastBatch(ctx context.Context, nextSeqHash [32]byte) error {
+	h := m.State.Height()
+	block, err := m.Store.LoadBlock(h)
+	if err != nil {
+		return fmt.Errorf("load block: height: %d: %w", h, err)
+	}
+
+	// check if the last block already produced with nextProposerHash set
+	if bytes.Equal(block.Header.NextSequencersHash[:], nextSeqHash[:]) {
+		m.logger.Debug("Last block already produced and applied.")
+	} else {
```

Comment:
if it's already produced, how do you know it's not already submitted?

Reply:
we got into this call flow when the hub expects final state update (`RotationInProgress`)

Reply:
not sure I got it

you can also go in there from here right?
https://github.com/dymensionxyz/dymint/blob/e1a51406345d238720987a4c5229b09c2d01a918/block/manager.go#L208

so how do you know it's not submitted already?

Reply:
after calling `CreateAndPostLastBatch`, the node will panic/change role.
so if this code runs, it means that **rotation in progress** on the hub, which means the last state update haven't been sent

Reply:
I see, so bytes.Equal(block.Header.NextSequencersHash[:], nextSeqHash[:])  will always be false unless this is after restart
👍